### DISCUSSION TITLE
When querying data, the result is intermittently returned as None.

### DISCUSSION THREAD

**Discussion by paytalab-y0ngha** [sqlalchemy/sqlalchemy#11146] [Usage Questions]
Hello,

We are operating a server using a combination of SQLAlchemy, Flask, and Gunicorn. We use PostgreSQL version 10 and SQLAlchemy version 1.3.22. We have a Master/Slave setup, with all basic read operations going through the Slave.

Lately, we keep encountering a problem where data that should exist is reported as missing. The issue is that it doesn't happen every day, but intermittently, making it hard for us to quickly identify the cause.

We have spent a lot of time trying to resolve this issue without success.

We create an Order object and proceed with db.session.add(...), db.session.flush(), db.session.commit(). (Let's say the primary key that can query this Order is '1234'.)

Then, we query the Order table for '1234'. We start the query with db.session.query(Order).filter(Order.identifier == '1234').

Here, the result is returned as None. Suspecting replication lag, we added a defensive logic to retry after 3 seconds, doing this a total of 4 times.

However, the result is None all 4 times.
In reality, when we searched the database with the identifier '1234' at that moment, we were able to confirm that the value existed as expected.

Has anyone experienced this issue before? We really want to solve this problem.

**Comment by CaselIT** [comment]
Hi,

Are you doing `db.session.query(Order).filter(Order.identifier == '1234')` in the same session that added the object? If not this may be due to the isolation level set on the DB? If you query from a session that's in a transaction with isolation level repeatable read or higher you would not see the new row until the transaction ends.

**Reply by paytalab-y0ngha** [reply]
Hi @CaselIT,

Thank you for showing interest in this discussion.

It's not the same session. It proceeds with a different session, and transactions are separated as well. The isolation level set in the DB is `READ COMMITTED`.

The API for committing and the API for reading are separated.

We have set the DEBUG log (Echo=True) for this issue and have secured the logs, but we don't consider there to be any meaningful logs.

**Reply by paytalab-y0ngha** [reply]
I posted the same question on Stackoverflow two months ago, but no one responded. The post on Stackoverflow mentioned a repeated query every 6 seconds for a duration of 24 seconds, while the current discussion mentions a repeated query every 3 seconds for a duration of 12 seconds. It would be good to recognize this as a change over the two months.

https://stackoverflow.com/questions/77784601/amazon-aurora-postgresql-returns-stale-data

**Reply by paytalab-y0ngha** [reply]
Here's some additional information to share.

We operate our service on a Kubernetes platform, and there are multiple Pods performing the Read operation. However, the issue occurs across all Pods.

If there's any information needed to facilitate communication on this issue, we are more than willing to provide prompt responses 😢

**Reply by paytalab-y0ngha** [reply]
However, it seems difficult to conclude that there's a problem with the database. I've looked at all the indicators of the database, and it was stable (and, to be fair, the services using SQLAlchemy were also stable). Crucially, if there had been an issue with the database, we would have faced failure in all queries, which was not the case.

**Reply by CaselIT** [reply]
thanks for the additional info.

I would try to rule out obvious things first:
- does running the select query in the same session before the committ (after the flush) and after committ work?
- does running the select query in the master db on a different session work?

This would rule out issues with types (example id is passed as string in the query but is an int in the db)

**Reply by zzzeek** [reply]
have you tried turning on SQL logging?    while verbose, if you put on debug-level logging it will show not just the queries but also the result rows returned.  

also when you do a "retry", if you are in the same transaction, that might be an issue depending on how this replication works.  you might want to try starting a new transaction for the retry.  or at least when you encounter the problem, open a brand new connection to PG right there that's separate from your usual session, run the SELECT and report if the row is actually there on an independent, new connection.

**Reply by paytalab-y0ngha** [reply]
Thank you for coming back @CaselIT ,

> does running the select query in the same session before the committ (after the flush) and after committ work?

No, it occurs in a different session.

> does running the select query in the master db on a different session work?

Today, I checked the logs that appear when queried from the master. However, we do not consider this to be due to replication delay because the replication delay time in the database has not increased, and it is generally expected to appear after retrying for the duration of 12 seconds as mentioned in the text.

The type of the ID(identifier) is in UUID format.

**Reply by paytalab-y0ngha** [reply]
Hi @zzzeek!

Yes, we have also turned on SQL logging. When attempted, we confirmed that the query was entered correctly, and the variables were also correctly inserted (here, the variable refers to the identifier value).

The retry process is conducted in an asynchronous manner and takes place in a different transaction. (To be precise, it occurs in a different Pod, Process. We are using Python's Celery approach. This means it's in a different session, different transaction, different process, different Pod...).

To identify the problem, we attempted the following:

We conducted a session.query(), and in cases where the result was None, we also conducted session logging in its entirety (wondering if perhaps the session was too old). However, that was not the case. Everything was fresh.

Additionally, we are using Datadog for monitoring, and we have confirmed that the actual number of rows returned for that query is zero.
(Of course, this was also confirmed in the logs.)

**Reply by paytalab-y0ngha** [reply]
We are using the following class.

```python
class RoutingSession(orm.Session):
  
    def __init__(self, db, autocommit=False, autoflush=False, **options):
        self.app = db.get_app()
        self.db = db
        self._name = None
        self._model_changes = {}
        orm.Session.__init__(
            self, autocommit=autocommit, autoflush=autoflush,
            bind=db.engine,
            binds=db.get_binds(self.app), **options)

    def get_bind(self, mapper=None, clause=None):

        try:
            state = get_state(self.app)
        except (AssertionError, AttributeError, TypeError) as err:
            current_app.logger.error("RoutingSession - cant get configuration. default bind. Error: {}".format(err))
            return orm.Session.get_bind(self, mapper, clause)

        # If there are no binds configured, use default SQLALCHEMY_DATABASE_URI
        if not state or not self.app.config['SQLALCHEMY_BINDS']:
            return orm.Session.get_bind(self, mapper, clause)

        # if want to user exact bind
        if self._name:
            if self._name == 'master':
                return orm.Session.get_bind(self, mapper, clause)
            
            return state.db.get_engine(self.app, bind=self._name)
        elif self._flushing:
            # current_app.logger.debug("RoutingSession - DATABASE MASTER ENGINE")
            return orm.Session.get_bind(self, mapper, clause)
        else:
            # if no bind is used connect to default
            # current_app.logger.debug("RoutingSession - DATABASE SLAVE ENGINE")
            return state.db.get_engine(self.app, bind='slave')

    def using_bind(self, name):
        s = RoutingSession(self.db)
        vars(s).update(vars(self))
        s._name = name
        return s


class RouteSQLAlchemy(SQLAlchemy):

    def __init__(self, *args, **kwargs):
        SQLAlchemy.__init__(self, *args, **kwargs)
        self.session.using_bind = lambda s: self.session().using_bind(s)

    def create_scoped_session(self, options=None):
        if options is None:
            options = {}
        scopefunc = options.pop('scopefunc', None)
        return orm.scoped_session(
            partial(RoutingSession, self, **options), scopefunc=scopefunc
        )

db = RouteSQLAlchemy()
```

When using actual queries, we use them as follows:

When binding to master: `db.session.using_bind("master").query(Order)`
When binding to slave: `db.session.query(Order)`

**Reply by zzzeek** [reply]
right so when the "unexpected None" thing happens, can you just immediately do a new query against "master" DB ?  just at the point at it happens, go into a mode where your program will poke around, use new connections with new transactions, etc. , see where the row is and where it isnt.

**Reply by paytalab-y0ngha** [reply]
@zzzeek 

When a None value was returned, we immediately bound the query to master and executed it for debugging purposes, only logging the results. At that time, there was a value.

However, we do not attribute this to replication lag. Since there has been no increase in the database replication time, we do not consider it to be due to replication lag. As mentioned in the text, it is generally expected to appear after retrying for 12 seconds.

**Reply by zzzeek** [reply]
im not attributing it to that either :)   it's just you are doing a query on your follower DB and the row isn't there, in the state of that transaction, at least.   so verifying 1. yes it's on the leader DB and 2. is it in a new transaction on the follower DB?  maybe our transaction is stale, that kind of thing.

**Reply by paytalab-y0ngha** [reply]
@zzzeek 

Ah, I see what you mean now. Thank you for explaining that so kindly once again.

The situation is that the data exists in the master DB but appears to be absent when checked in the slave DB. Could it be due to long-running transactions?
Hmm... would it be helpful if I sent you the logs below?

```
ERROR:console:pre_read_order - NoResultFound(), order_identifier: b8224c41-9d85-425c-bd13-6665c7458af5
DEBUG:sqlalchemy.pool.impl.QueuePool:Connection <Psycopg2TracedConnection at 0xffff7ea0a900 for psycopg2.extensions.connection at 0xffff7ebf92c0> checked out from pool
DEBUG:flask.app:[SQLAlchemy] Transaction is starting on a connection that has been in use for 0:21:40.763573. (connection pid: 20914, established at: 2024-03-12 08:24:37.532427)
```

**Reply by zzzeek** [reply]
There's nothing I can do in this end with logs.   If the data was definitely replicated, you confirm that by a. making a brand new connection at that moment and confirming the row is accessible via your DBAPI etc  b. checking on your current connection, seeing if ending the transaction and starting a new one is the issue, Id even go so far as to say emit a raw "ROLLBACK" SQL command, then look into different levels of [DISCARD](https://www.postgresql.org/docs/current/sql-discard.html) to see if any of that is causing the state to be stale.

**Reply by paytalab-y0ngha** [reply]
@zzzeek 

Thank you for getting back to me so quickly with the information.

Please note that I am currently based in South Korea, so it may be difficult for me to respond quickly.

I'll discuss what you've mentioned with the people I'm working with and get back to you around 01:00 UTC.

Thank you so much for your quick response.

**Reply by paytalab-y0ngha** [reply]
Hi @zzzeek,

Morning has dawned in Korea Time. I'm back.

We've assessed that it might be difficult to test the suggested method in a live operational environment. While executing a ROLLBACK seems feasible since it's at the transaction level, DISCARD could be risky as it operates at the DB level, potentially deleting all query plans and significantly impacting our service.

Is there a way in SQLAlchemy to operate DISCARD for only one session? We haven't been able to find documentation on this.

We plan to continue researching to solve this issue today.

**Reply by zzzeek** [reply]
yes when you have a session, it refers to a single conneciton.   you can get at it like this:

```
conn = session.connection()
conn.exec_driver_sql("DISCARD <thing>")
```

you can also have that connection be [detached](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.detach) from the pool so it will be closed / garbage collected when you close() it

```
conn.detach()
```

**Reply by paytalab-y0ngha** [reply]
Oh, you were still active!! 😂 @zzzeek 

Would it be a good idea to execute `DISCARD ALL` at least once?

Thank you for kindly explaining with code.

Regarding the issue you inquired about yesterday, we have confirmed that in our system, a Rollback is performed at the end of every Read transaction. Therefore, it is a case where a new Transaction is being created!

**Reply by zzzeek** [reply]
i dont know!  i dont know why your query returns None.   I'd definitely make sure a SELECT is being run and look at what the rows are.

**Reply by paytalab-y0ngha** [reply]
We have decided to add logs in the following manner:

After data is added via the add() function in the DB, if a None Exception occurs when querying this data from another API request (i.e., a different transaction), we will attempt to query again directly from the Slave.
After that, we will use DISCARD ALL to reset the session, and try querying once more from the Slave.
Furthermore, we will query from the Master and record the log.

I also truly think this issue is nonsensical.

![스크린샷 2024-03-13 오전 11 17 24](https://github.com/sqlalchemy/sqlalchemy/assets/100120153/9ee82d65-a42f-4a53-a1e4-fc9c29c3ecce)

The photo above shows that we actually executed the query and that no results were returned.

```
(KST) Mar 13 10:18:18.750 i-0685e8d18e981d91d api-server create_order - Insert Success: {'identifier': UUID('aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee' ...)}
(KST) Mar 13 10:18:18.751 i-0685e8d18e981d91d api-server verify_order - Update Success: aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee

(KST) Mar 13 10:18:21.159 i-0685e8d18e981d91d celery-worker publish_order - try 1/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:21.159 i-0685e8d18e981d91d celery-worker [2024-03-13 01:18:20,978: INFO/ForkPoolWorker-4] publish_order - try 1/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:21.159 i-0685e8d18e981d91d celery-worker NoResultFound Error Excepted caller=publish_order, order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None
(KST) Mar 13 10:18:21.159 i-0685e8d18e981d91d celery-worker [2024-03-13 01:18:20,985: INFO/ForkPoolWorker-4] NoResultFound Error Excepted order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None

(KST) Mar 13 10:18:23.709 i-0231309ac319e0c0d celery-worker publish_order - try 2/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:23.709 i-0231309ac319e0c0d celery-worker [2024-03-13 01:18:23,147: INFO/ForkPoolWorker-5] publish_order - try 2/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:23.709 i-0231309ac319e0c0d celery-worker NoResultFound Error Excepted caller=publish_order, order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None
(KST) Mar 13 10:18:23.709 i-0231309ac319e0c0d celery-worker [2024-03-13 01:18:23,156: INFO/ForkPoolWorker-5] NoResultFound Error Excepted caller=publish_order, order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None

(KST) Mar 13 10:18:25.710 i-0231309ac319e0c0d celery-worker publish_order - try 3/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:25.710 i-0231309ac319e0c0d celery-worker [2024-03-13 01:18:25,219: INFO/ForkPoolWorker-1] publish_order - try 3/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:25.710 i-0231309ac319e0c0d celery-worker NoResultFound Error Excepted caller=publish_order, order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None
(KST) Mar 13 10:18:25.710 i-0231309ac319e0c0d celery-worker [2024-03-13 01:18:25,228: INFO/ForkPoolWorker-1] NoResultFound Error Excepted caller=publish_order, order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None

(KST) Mar 13 10:18:28.262 i-0685e8d18e981d91d celery-worker publish_order - try 4/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:28.263 i-0685e8d18e981d91d celery-worker [2024-03-13 01:18:27,280: INFO/ForkPoolWorker-4] publish_order - try 4/4, order_identifier -> aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status -> 1
(KST) Mar 13 10:18:28.263 i-0685e8d18e981d91d celery-worker NoResultFound Error Excepted caller=publish_order, order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None
(KST) Mar 13 10:18:28.263 i-0685e8d18e981d91d celery-worker [2024-03-13 01:18:27,287: INFO/ForkPoolWorker-4] NoResultFound Error Excepted caller=publish_order, order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, result=None
(KST) Mar 13 10:18:28.263 i-0685e8d18e981d91d celery-worker publish_order - The recursive call was executed four times, but it terminated due to meeting an 'if' condition. order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status=1
(KST) Mar 13 10:18:28.263 i-0685e8d18e981d91d celery-worker [2024-03-13 01:18:27,287: ERROR/ForkPoolWorker-4] publish_order - The recursive call was executed four times, but it terminated due to meeting an 'if' condition. order_identifier=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee, order_status=1
```

**Reply by paytalab-y0ngha** [reply]
It means that if we get a `None` result on the first query, we will open an additional session to try querying again. After that, we will proceed with `DISCARD` and then execute the query once more in the same session.

**Reply by paytalab-y0ngha** [reply]
We are planning to deploy this content now and wait for the logs to be collected.

Thank you for your help today. @zzzeek

**Reply by paytalab-y0ngha** [reply]
We have added logs and deployed them, but the logs are not being collected. Since this issue occurs intermittently, we will leave a note again once the symptom is confirmed.

Thank you.
